### PR TITLE
APIM enhancements

### PR DESCRIPTION
**Pull request by Aboli-msft**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 1]
```diff
@@ -337,6 +337,19 @@ class APIManagement: AzSVTBase
     {
 		if(($null -ne $this.APIMContext) -and ($null -ne $this.APIMAPIs))
 		{
+			if ( ($this.APIMAPIs | Measure-Object).Count -gt $this.ControlSettings.APIManagement.MaxAllowedAPICount)
+			{
+				$scanSource = [RemoteReportHelper]::GetScanSource();
+				if($scanSource -eq [ScanSource]::Runbook)
+				{
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: We can instead use feature flight flag here. Feature flight allows you to control scan source.

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 2]
```diff
@@ -337,6 +337,19 @@ class APIManagement: AzSVTBase
     {
 		if(($null -ne $this.APIMContext) -and ($null -ne $this.APIMAPIs))
 		{
+			if ( ($this.APIMAPIs | Measure-Object).Count -gt $this.ControlSettings.APIManagement.MaxAllowedAPICount)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: If MaxAllowedAPICount is greater than total # of APIs, and scan source is SDL. What would be the status of this control? Would it be Manual?

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 3]
```diff
@@ -390,7 +404,7 @@ class APIManagement: AzSVTBase
 				# If number of APIs is higher than MaxAllowedAPICount, the control is marked as Verify.
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: Shouldn't there be a check here to for scan source 'CA'?

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 4]
```diff
@@ -494,12 +508,27 @@ class APIManagement: AzSVTBase
     {
 		if ( $null -ne $this.APIMContext)
 		{
+			if ( ($this.APIMAPIs | Measure-Object).Count -gt $this.ControlSettings.APIManagement.MaxAllowedAPICount)
+			{
+				$scanSource = [RemoteReportHelper]::GetScanSource();
+				if($scanSource -eq [ScanSource]::Runbook)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: 1. Feature flight can be used here instead to check scan source.
2. If scan source in SDL, else condition should get executed.

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 5]
```diff
@@ -867,7 +897,21 @@ class APIManagement: AzSVTBase
 		$JWTValidatePolicyNotFound = @()
 		if(($this.APIUserAuth -ne 'ResourceNotFound') -and ($null -ne $this.APIMContext) -and ($null -ne $this.APIMAPIs))
 		{
-			$this.APIMAPIs | ForEach-Object {
+			if ( ($this.APIMAPIs | Measure-Object).Count -gt $this.ControlSettings.APIManagement.MaxAllowedAPICount)
+			{
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: 1. Feature flight can be used here instead to check scan source.
2. If scan source in SDL, else condition should get executed.

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 (hunk 3 above)
Comment (Aboli-msft) [reply src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: Added check for CA

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 (hunk 5 above)
Comment (Aboli-msft) [reply src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: Added featureFlight and corrected if condition

File: src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1 [hunk 6]
```diff
@@ -337,6 +337,17 @@ class APIManagement: AzSVTBase
...
+			if ($isExcludeControlForLargeAPIM -and (($this.APIMAPIs | Measure-Object).Count -gt $this.ControlSettings.APIManagement.MaxAllowedAPICount))
+			{
+				# If number of APIs is higher than MaxAllowedAPICount, the control is marked as Verify in CA mode.
+				# This check has been implemented because of the execution time of control as well as socket and memory limit exhaustion in case of APIMs with large number of APIs
+				# In this case, user must follow the FAQ provided in recommendation to check API/Operation level policy
+					$controlResult.CurrentSessionContext.Permissions.HasRequiredAccess = $false;
+					$controlResult.AddMessage([VerificationResult]::Verify,[MessageData]::new("Total API count: $(($this.APIMAPIs | Measure-Object).Count)`n`rPlease use client certificates to secure access to the back-end service of an API and protects data in transit from network layer.To enable client certificate authentication from Azure portal please refer https://docs.microsoft.com/en-us/azure/api-management/api-management-howto-mutual-certificates and https://docs.microsoft.com/en-us/azure/api-management/api-management-authentication-policies."));
```
Comment (v-zhgup) [inline src/AzSK/Framework/Core/SVT/Services/APIManagement.ps1]: please get this message reviewed. It should be like "Total API count for the current instance exceeded maximum limit, please verify manually using ...... provide the az commands that would be helpful for end users to verify: $(($this.APIMAPIs | Measure-Object).Count)

**Pull request merged**